### PR TITLE
fix issues when cleanupTxn called multiple times

### LINKED ISSUE #1823: `(*TxnCoordSender).cleanupTxn` is sometimes called twice and deadlocks
`make CPUS=4,4,4,4,1,2,3,5,6,7,8,9,10 TESTFLAGS='--verbosity=2' PKG=./kv TESTTIMEOUT=5m TESTS=TestTxnDB`

Run that a few times and you'll eventually get https://gist.github.com/65836cb3f0c0e9ca3776

### PR DESCRIPTION
**Pull request by tbg**
now the txnMeta.txnEnd channel is set to nil by whoever gets
there first, everyone after is a no-op. Fixed a small race
in heartbeats regarding that change.

fixes #1823.

### CODE REVIEW

**Comment by MBerhault** [comment]
LGTM

File: kv/txn_coord_sender_test.go [hunk 1]
```diff
@@ -746,3 +746,41 @@ func TestTxnDrainingNode(t *testing.T) {
...
+// TestTxnCoordIdempotentCleanup verifies that cleanupTxn is idempotent.
+func TestTxnCoordIdempotentCleanup(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	s := createTestDB(t)
+	defer s.Stop()
+
+	txn := newTxn(s.Clock, proto.Key("a"))
+	pReply := &proto.PutResponse{}
+	key := proto.Key("a")
+	call := proto.Call{
```
Comment (tamird) [inline kv/txn_coord_sender_test.go]: `proto.PutCall()` might help you here

Comment (tamird) [inline kv/txn_coord_sender_test.go]: oh, transaction. disregard

File: kv/txn_coord_sender_test.go [hunk 2]
```diff
@@ -746,3 +746,41 @@ func TestTxnDrainingNode(t *testing.T) {
...
+		Args:  createPutRequest(key, []byte("value"), txn),
+		Reply: pReply,
+	}
+	if err := sendCall(s.Sender, call); err != nil {
+		t.Fatal(err)
+	}
+
+	if pReply.GoError() != nil {
```
Comment (tamird) [inline kv/txn_coord_sender_test.go]: should check `.Error` for symmetry with the below

**Comment by tamird** [comment]
LGTM

**Comment by bdarnell** [comment]
LGTM

File: kv/txn_coord_sender_test.go (hunk 2 above)
Comment (tschottdorf) [inline kv/txn_coord_sender_test.go]: done.

File: kv/txn_coord_sender.go [hunk 3]
```diff
@@ -694,7 +695,7 @@ func (tc *TxnCoordSender) cleanupTxn(trace *tracer.Trace, txn proto.Transaction,
 	// Trigger intent resolution and heartbeat shutdown.
 	trace.Event("coordinator stops")
 	txnMeta.txnEnd <- resolved // buffered, so does not block
```
Comment (tamird) [inline kv/txn_coord_sender.go]: this comment could use a correction since setting to nil makes it block :)

Comment (tschottdorf) [inline kv/txn_coord_sender.go]: done.

**Pull request merged**